docs/contribute/saros-server.md

```diff
@@ -0,0 +1,40 @@
+---
+title: Saros Server Development
+---
+
+## What is the Saros Server
+
+Please refer to the [User Documentation](../documentation/saros-server.md) for an explaination and the use-case of the saros server.
+
+Technically it is a headless session-host.
+
+## Code Structure
+
+The Saros Server codebase is comparably slim, as it only requires a [small subset of components](https://github.com/saros-project/saros/blob/master/server/src/saros/server/ServerContextFactory.java) compared to IDE-Plugins. Therefor it may also help to understand Saros overall architecture and as a reference for starting new Saros-Implementations.
+
+Notable exceptions are the [filesystem](https://github.com/saros-project/saros/tree/master/server/src/saros/server/filesystem), [editor](https://github.com/saros-project/saros/tree/master/server/src/saros/server/editor) and [preferences](https://github.com/saros-project/saros/tree/master/server/src/saros/server/preferences) implementations, which are usually adapted from IDE-specific APIs.
+
+## Open Topics
+
+### Permission System
+
+Traditionally the Session-Host is responsible for moderating an active Saros-Session. The Host-Role allows adding new users and accepting new projects into the session. A headless implementation like the server cannot do this, which is why the server currectly accepts every request for invitations and new projects. Therefor it can only be executed in a trusted environment.
+
+The server will need some kind of permission system to allow a remote user to manage these tasks. The simplest (and likely sufficient) variant would move this role to some kind of "Admin"-User, that may be configured for a given server.
+
+#### Challenges
+
+- New Activities will likely need to be introduced, that need to be understood by all existing IDE-Clients and show the appropriate dialogs.
+- There needs to be a system to transfer the admin rights, otherwise the admin-user becomes the new session-host, that will leave behind a broken (rather then disconnected) session, once he leaves.
+
+### Super-Server
+
+One server currently represents only one session. To lift this short-coming a Super-Server could be developed, that starts a new session-server on demand. This would also play nicely with the idea of an admin-user, as the user requesting the session could automatically be configured to be the first admin.
```

Comment:
This does not scale well (consumes to much memory) if you fork a child process for each session. I think 64 MB is the minimum of memory (guess this is still to low) that the process would need, So hosting about 30 session will already consume almost 2 GB.

Reply:
This idea exists for quite some time already and has some nice properties:

- It is nicely capsulated, which makes this a good issue for a newcomer to the project (e.g. bachelor thesis)
- One session crashing would not crash the entire server
- It would likely simplify the code

Of course, each separate process and JVM would need additional memory, but a connected saros-server only takes about 28mb on my system, which already includes a good chunk of saros objects. So you are overestimating, what resources the JVM needs these days.

My guess is, that most of the memory would be used by the open documents / jupiter-algorithm / etc anyway, as saros was not designed to be particular memory-efficient anyway.

As a result, the server would need quite a lot of memory to be able to handle 30 sessions anyway. I expect an early release of this feature to be used for 5-10 session max. I think the main benefit is the ease of use, that it provides when you do not need to start the server manually.

That said, this is clearly something that needs to be evaluated by whoever want to implement this feature. I will this to the docs.